<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек-и-загрузка-данных" data-toc-modified-id="Импорт-библиотек-и-загрузка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотек и загрузка данных</a></span></li><li><span><a href="#Анализ-и-предварительная-обработка-данных" data-toc-modified-id="Анализ-и-предварительная-обработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Анализ и предварительная обработка данных</a></span></li><li><span><a href="#Выводы:" data-toc-modified-id="Выводы:-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Выводы:</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Оцениваем-качество-предсказаний-модели-на-исходных-данных" data-toc-modified-id="Оцениваем-качество-предсказаний-модели-на-исходных-данных-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Оцениваем качество предсказаний модели на исходных данных</a></span></li><li><span><a href="#Оцениваем-качество-предсказаний-модели-на-модернизированных-данных" data-toc-modified-id="Оцениваем-качество-предсказаний-модели-на-модернизированных-данных-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Оцениваем качество предсказаний модели на модернизированных данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

В рамках проекта нам нужно защитить данные клиентов страховой компании «Хоть потоп». Нужно разработать такой метод преобразования данных, чтобы по полученныму результату было сложно восстановить персональную информацию. Корректность его работы необходимо обосновать.  
Нужно защитить данные, преобразовав их так, чтобы в последствии качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.  

**План по выполнению проекта**  
1. Загрузить и изучить данные;  
2. Ответьте на вопрос и обоснуйте решение:   
   Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново).  
      a. Изменится. Приведите примеры матриц;  
      b. Не изменится. Укажите, как связаны параметры линейной регрессии в исходной задаче и в преобразованной.  
      
3. Предложить алгоритм преобразования данных для решения задачи. Обосновать, почему качество линейной регрессии не поменяется.  
4. Запрограммировать этот алгоритм, применив матричные операции. Проверить, что качество линейной регрессии из sklearn не отличается до и после преобразования. Применить метрику R2.  

**Описание данных**  
Набор данных находится в файле /datasets/insurance.csv.  
*Признаки:* пол, возраст и зарплата застрахованного, количество членов его семьи.  
*Целевой признак:* количество страховых выплат клиенту за последние 5 лет.  

## Загрузка данных

### Импорт библиотек и загрузка данных

In [1]:
# import
import pandas as pd
import numpy as np
import os

from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_columns', None)

In [2]:
# подгружаем данные
pth0_1 = '/datasets/insurance.csv'
pth0_2 = 'C:/Users/username/Desktop/insurance.csv'

if os.path.exists(pth0_1):
    data = pd.read_csv(pth0_1)
elif os.path.exists(pth0_2):
    data = pd.read_csv(pth0_2)
else:
    print('Check the PATH to data file please')

### Анализ и предварительная обработка данных

In [3]:
# выводим первичную информацию о наборе данных
print(data.head())
print(data.info())
print(data.describe())

   Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB
None
               Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.0000

Очевидно, что кол-во членов семьи и возраст должны быть целочисленным типом данных, приведём к нему.

In [4]:
data[['Зарплата', 'Возраст']] = data[['Зарплата', 'Возраст']].astype('int')
#data.info()

Далее проверим на явные дубликаты и уникальные значения некоторые столбцы

In [5]:
# проверяем на явные дубликаты
data.duplicated().sum()

153

In [6]:
columns = ['Пол', 'Члены семьи', 'Страховые выплаты']
for column in columns:
    print('Столбец:', column, sep=' ')
    print('Кол-во уникальных значений:', len(data[column].unique()), sep=' ')
    print('Уникальные значения:', sorted(data[column].unique()), sep='\n')
    print()

Столбец: Пол
Кол-во уникальных значений: 2
Уникальные значения:
[0, 1]

Столбец: Члены семьи
Кол-во уникальных значений: 7
Уникальные значения:
[0, 1, 2, 3, 4, 5, 6]

Столбец: Страховые выплаты
Кол-во уникальных значений: 6
Уникальные значения:
[0, 1, 2, 3, 4, 5]



### Выводы:  
В целом, данные не требуют дополнительной подготовки.  
Были обнаружены 153 полных дубликатов строк, с учётом 5000 строк, заполненных только цифрами, такое вполне могло случиться. Поэтому было принято решение оставить их как есть.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопрос:**  
Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново).

**Ответ:**  
Не изменятся.

**Обоснование:**

Запишем новое представление матрицы признаков с учетом домнажения на обратную матрицу:

$$
Z = XP
$$

Тогда формулы c учетом новых введений изменят свой внешний вид.  
Предсказания:

$$
A = ZW
$$

Формула обучения:

$$
W = (Z^T Z)^{-1} Z^T y,
$$

где W - вектор весов

Получается, что выражение для $W$ преобразуется как:

$$W = (Z^TZ)^{-1} Z^Ty =$$
$$= ((XP)^T(XP))^{-1} (XP)^Ty =$$
$$= P^{-1}((XP)^{T}X)^{-1} (XP)^Ty =$$
$$= P^{-1}((XP)^{T}X)^{-1} P^TX^Ty =$$
$$= P^{-1}(P^TX^TX)^{-1} P^TX^Ty =$$
$$= P^{-1}(X^TX)^{-1}(P^T)^{-1} P^TX^Ty$$

Далее сокращаем:

$$(P^{T})^{-1} P^T = P^{-1}(X^TX)^{-1} X^Ty$$

Произведем замену

$$W = (X^TX)^{-1}X^Ty = P^{-1}w$$

Подставим для $Z$ и $W$ в формулу для расчета $A$:

$$A=ZW = XPP^{-1}w = Xw = a$$

Выражения тождественны. Векторы предсказаний совпали.

### Вывод

Исходная формула вектора предсказаний равна формуле вектора предсказаний в случае, для которого матрицу признаков умножили на обратимую матрицу. Качество линейной регрессии не изменится.

Рекомендую к ознакомлению статью на Хабр: https://habr.com/ru/articles/479398/

## Алгоритм преобразования

**Алгоритм**

Алгоритм составляем на основании доказательства приведённого разделом выше: матрицу признаков необходимо умножить на обратимую матрицу той же размерности, все значения которой сгенерированы случайно.

**Обоснование**

Для обоснования проведём небольшие практические расчёты.

In [7]:
# разделим пространство признаков и целевой признак
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

После удаления целевого признака из данных, осталось ровно 4 колонки, следовательно, матрицу с признаками мы будем домнажать на матрицу размером (4, 4), сгенерированную случайно.  
**Важно!** Нужно проверить, что у случайно сгенерированной матрицы есть обратная матрица.

In [8]:
# создадим рандомную матрицу нужного размера
rndm_mtx = np.random.randn(4, 4)

# проверяем её на существование обратной матрицы
try:
    inv_mtx = np.linalg.inv(rndm_mtx)
    print('\nОбратная матрица к заданной существует')
except:
    print('\nОбратная матрица к заданной не существует')


Обратная матрица к заданной существует


Далее проводим рассчёты согласно формулам из предыдущего раздела:

$$
a = Xw
$$

$$
w = (X^T X)^{-1} X^T y
$$

In [9]:
# для исходных признаков
w = np.linalg.inv(features.T.dot(features)).dot(features.T).dot(target)
a = features @ w

In [10]:
# преобразуем матрицу признкаов и получим предсказания для неё
features_upg = features.dot(rndm_mtx)
w2 = np.linalg.inv(features_upg.T.dot(features_upg)).dot(features_upg.T).dot(target)
a2 = features_upg @ w2

In [11]:
# разница между предсказаниями составляет
print((a - a2).sum())

4.0468440228882095e-06


Очевидно, что разница на столько мала, что может быть признана несущественной.

### Выводы  
Согласно произведённым практическим расчётам видно, что разница между предсказаниями на основании исходной матрицы признаков и изменённой, полученной за счёт домножения исходной матрицы на матрицу из случайных значений, но обладающую обратной матрицей, несущественена, а следовательно, алгоритм позволяет изменять исходные признаки без снижения качества предсказаний модели.

## Проверка алгоритма

### Оцениваем качество предсказаний модели на исходных данных

In [23]:
# создаем обучающую и тестовую выборки
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.25, random_state=12345)

In [24]:
# модель с исходными признаками без преобразования
model = LinearRegression()
model.fit(features_train, target_train)
r2_lr_of = r2_score(target_test, model.predict(features_test)) # original features
print("R2 =", r2_lr_of)

R2 = 0.4352275684083322


In [25]:
# добавим масштабирование признаков
pipeline = Pipeline([("standard_scaller", StandardScaler()),("linear_regression", LinearRegression())])
pipeline.fit(features_train, target_train)
r2_lr_sof = r2_score(target_test, pipeline.predict(features_test)) # scaled original features
print("R2 =", r2_lr_sof)

R2 = 0.43522756840833376


In [26]:
# разница
print('Разница между предсказаниями на масштабированных и немасштабированных оригинальных данных составляет:', 
      r2_lr_of - r2_lr_sof, sep='\n')

Разница между предсказаниями на масштабированных и немасштабированных оригинальных данных составляет:
-1.5543122344752192e-15


Можно считать разницу несущественной.

###  Оцениваем качество предсказаний модели на модернизированных данных

In [27]:
# посмотрим на полученную ранее случайную матрицу
print(rndm_mtx)

[[-0.68924236 -0.63186005 -0.39397444 -0.84142046]
 [-1.35702699 -0.63104004 -0.92390933  1.31619301]
 [-0.38079073 -0.00416371 -0.14514724 -0.28838423]
 [-0.52483315 -0.75832429  0.2402151   1.06256136]]


In [28]:
# посмотрим на результат преобразования матрицы принаков
display(features.head(5))
display(features_upg.head(5))

Пол  Возраст  Зарплата  Члены семьи
0    1       41     49600            1
1    0       46     38000            1
2    0       29     21000            0
3    0       21     41700            2
4    1       28     26100            0

0           1            2             3
0 -18944.072240 -233.783008 -7237.337253 -14249.672768
1 -14532.995699 -188.007273 -5557.854816 -10896.993311
2  -8035.959049 -105.738142 -3074.885456  -6017.899239
3 -15908.520548 -188.395336 -6071.561664 -11995.857227
4  -9977.323972 -126.973900 -3814.606456  -7490.816427

In [29]:
# разбиваем пространство преобразованных признаков на обучающую и тестовую выборки
features_train_upg, features_test_upg, target_train, target_test = train_test_split(features_upg, target, test_size=0.25, random_state=12345)

In [31]:
# модель с модернизированными признаками без преобразования
model = LinearRegression()
model.fit(features_train_upg, target_train)
r2_lr_uf = r2_score(target_test, model.predict(features_test_upg)) # upgrated features
print("R2 =", r2_lr_uf)

R2 = 0.43522756840825605


In [33]:
# добавим масштабирование модернизированных признаков
pipeline = Pipeline([("standard_scaller", StandardScaler()),("linear_regression", LinearRegression())])
pipeline.fit(features_train_upg, target_train)
r2_lr_suf = r2_score(target_test, pipeline.predict(features_test_upg)) # scaled upgrated features
print("R2 =", r2_lr_suf)

R2 = 0.43522756840835064


In [34]:
# итого, разница между результатами модели на оргинальных данных и преобразованных составит
print('Разница составляет:', r2_lr_sof - r2_lr_suf)

Разница составляет: -1.687538997430238e-14


In [35]:
summary = pd.DataFrame(data= [r2_lr_of,
                      r2_lr_sof,
                      r2_lr_uf,
                      r2_lr_suf], 
                      columns=['R2'], 
                      index=['Качество на исходных данных',
                            'Качество на исходных данных c масштабированием',
                            'Качество на преобразованных признаках',
                            'ЛКачество на преобразованных признаках с масштабированием',])
summary

R2
Качество на исходных данных                         0.435228
Качество на исходных данных c масштабированием      0.435228
Качество на преобразованных признаках               0.435228
ЛКачество на преобразованных признаках с масшта...  0.435228

### Вывод  
Результаты метрики R2, полученные для оригинальной матрицы признаков и преобразованной матрицы практически совпадают, что подтверждает верность теоретического обоснования из второго раздела проекта, а также предварительную оценку испльзования алгоритма из третьего.

## Общий вывод  
1. Предварительный анализ исходных данных показал их адекватность, мы изменили только тип данных для двух колонок. Также были обнаружено явные дубликаты, но в виду их небольшого количества и условия, все строки представляют собой наборы цифр, было принято решение оставить их;  
2. Ответили на вопрос: "Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?" Не изменится, привели теоретическое обоснование ответа;  
3. Предложили алгоритм сокрытия данных клиентов за счёт умножения матрицы признаков на обратимую матрицу, наполненную случайными величинами. Провели предварительный расчёт;  
4. Провели сравнение результатов работы линейной регрессии на оригинальном пространстве признаков и преобразованном, согласно предложенному алгоритму. **Исследовали метрику R2, качество работы модели не изменилось.**